File: src/main/java/ltdjms/discord/aichat/domain/AIServiceConfig.java
package ltdjms.discord.aichat.domain;

import ltdjms.discord.shared.DomainError;
import ltdjms.discord.shared.EnvironmentConfig;
import ltdjms.discord.shared.Result;
import ltdjms.discord.shared.Unit;

/**
 * AI 服務配置，包含連線資訊與參數。
 *
 * @param baseUrl AI 服務 Base URL (例如: https://api.openai.com/v1)
 * @param apiKey API 金鑰
 * @param model 模型名稱 (例如: gpt-3.5-turbo)
 * @param temperature 溫度 (0.0-2.0，控制回應隨機性)
 * @param timeoutSeconds 連線逾時秒數（僅用於建立連線，不限制推理時間）
 * @param showReasoning 是否顯示推理內容（預設: false）
 * @param enableMarkdownValidation 是否啟用 Markdown 格式驗證（預設: true）
 * @param streamingBypassValidation 串流模式是否跳過驗證（預設: false）
 * @param maxMarkdownValidationRetries Markdown 驗證最大重試次數（已停用，保留相容性）
 * @param enableAutoFix 是否啟用 Markdown 自動修復（已停用，保留相容性）
 */
public record AIServiceConfig(
    String baseUrl,
    String apiKey,
    String model,
    double temperature,
    int timeoutSeconds,
    boolean showReasoning,
    boolean enableMarkdownValidation,
    boolean streamingBypassValidation,
    int maxMarkdownValidationRetries,
    boolean enableAutoFix) {

  /** 從 EnvironmentConfig 建立配置。 */
  public static AIServiceConfig from(EnvironmentConfig env) {
    return new AIServiceConfig(
        env.getAIServiceBaseUrl(),
        env.getAIServiceApiKey(),
        env.getAIServiceModel(),
        env.getAIServiceTemperature(),
        env.getAIServiceTimeoutSeconds(),
        env.getAIShowReasoning(),
        env.getAIMarkdownValidationEnabled(),
        env.getAIMarkdownValidationStreamingBypass(),
        0,
        false);
  }

  /** 驗證配置。 */
  public Result<Unit, DomainError> validate() {
    if (baseUrl == null || baseUrl.isBlank()) {
      return Result.err(DomainError.invalidInput("AI_SERVICE_BASE_URL cannot be empty"));
    }
    if (apiKey == null || apiKey.isBlank()) {
      return Result.err(DomainError.invalidInput("AI_SERVICE_API_KEY cannot be empty"));
    }
    if (model == null || model.isBlank()) {
      return Result.err(DomainError.invalidInput("AI_SERVICE_MODEL cannot be empty"));
    }
    if (temperature < 0.0 || temperature > 2.0) {
      return Result.err(
          DomainError.invalidInput("AI_SERVICE_TEMPERATURE must be between 0.0 and 2.0"));
    }
    if (timeoutSeconds < 1 || timeoutSeconds > 120) {
      return Result.err(
          DomainError.invalidInput("AI_SERVICE_TIMEOUT_SECONDS must be between 1 and 120"));
    }
    return Result.okVoid();
  }
}

File: packages/ai/src/config/ai-service-config.ts
import { z } from 'zod';
import { DomainError, ok, okVoid, err, type Result } from '@ltdjms/shared';
import type { EnvironmentConfig } from '@ltdjms/shared';

/**
 * Zod schema for AI service configuration values.
 * Matches Java AIServiceConfig record fields.
 */
export const AIServiceConfigSchema = z.object({
  baseUrl: z.string().min(1, 'AI service base URL is required'),
  apiKey: z.string().min(1, 'AI service API key is required'),
  model: z.string().min(1, 'AI model name is required'),
  temperature: z.number().min(0).max(2),
  timeoutSeconds: z.number().int().min(1).max(120),
  showReasoning: z.boolean().default(false),
  enableMarkdownValidation: z.boolean().default(true),
  streamingBypassValidation: z.boolean().default(false),
});

export type AIServiceConfigValues = z.infer<typeof AIServiceConfigSchema>;

/**
 * AI service configuration — immutable value object.
 * Constructed from EnvironmentConfig or from a raw values object.
 * Matches Java AIServiceConfig record.
 */
export class AIServiceConfig {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly model: string;
  readonly temperature: number;
  readonly timeoutSeconds: number;
  readonly showReasoning: boolean;
  readonly enableMarkdownValidation: boolean;
  readonly streamingBypassValidation: boolean;
  readonly enableThinking: boolean;

  private constructor(values: AIServiceConfigValues) {
    this.baseUrl = values.baseUrl;
    this.apiKey = values.apiKey;
    this.model = values.model;
    this.temperature = values.temperature;
    this.timeoutSeconds = values.timeoutSeconds;
    this.showReasoning = values.showReasoning;
    this.enableMarkdownValidation = values.enableMarkdownValidation;
    this.streamingBypassValidation = values.streamingBypassValidation;
    // DeepSeek models automatically enable thinking
    this.enableThinking = values.showReasoning || /deepseek/i.test(values.model);
  }

  /**
   * Creates an AIServiceConfig from the shared EnvironmentConfig.
   */
  static from(env: EnvironmentConfig): AIServiceConfig {
    return new AIServiceConfig({
      baseUrl: env.getAIServiceBaseUrl(),
      apiKey: env.getAIServiceApiKey(),
      model: env.getAIServiceModel(),
      temperature: env.getAIServiceTemperature(),
      timeoutSeconds: env.getAIServiceTimeoutSeconds(),
      showReasoning: env.getAIShowReasoning(),
      enableMarkdownValidation: env.getAIMarkdownValidationEnabled(),
      streamingBypassValidation: env.getAIMarkdownValidationStreamingBypass(),
    });
  }

  /**
   * Creates an AIServiceConfig from a raw values object (for testing).
   */
  static fromValues(values: Partial<AIServiceConfigValues> & {
    baseUrl: string;
    apiKey: string;
    model: string;
  }): AIServiceConfig {
    const parsed = AIServiceConfigSchema.parse({
      temperature: 0.7,
      timeoutSeconds: 30,
      showReasoning: false,
      enableMarkdownValidation: true,
      streamingBypassValidation: false,
      ...values,
    });
    return new AIServiceConfig(parsed);
  }

  /**
   * Validates range constraints.
   * Returns okVoid or err(DomainError).
   */
  validate(): Result<void, DomainError> {
    const result = AIServiceConfigSchema.safeParse({
      baseUrl: this.baseUrl,
      apiKey: this.apiKey,
      model: this.model,
      temperature: this.temperature,
      timeoutSeconds: this.timeoutSeconds,
      showReasoning: this.showReasoning,
      enableMarkdownValidation: this.enableMarkdownValidation,
      streamingBypassValidation: this.streamingBypassValidation,
    });

    if (!result.success) {
      const firstIssue = result.error.issues[0];
      return err(DomainError.invalidInput(
        `AI config validation failed: ${firstIssue.path.join('.')} — ${firstIssue.message}`,
      ));
    }

    return okVoid<DomainError>() as unknown as Result<void, DomainError>;
  }

  /** DeepSeek model detection. */
  isDeepSeekModel(): boolean {
    return /deepseek/i.test(this.model);
  }
}
